miden-crypto/src/dsa/eddsa_25519/mod.rs

```diff
@@ -0,0 +1,191 @@
+        let shared = self.to_x25519().diffie_hellman(&pk_e.inner);
+        SharedSecret::new(shared)
+    }
+
+    /// Converts this Ed25519 secret key into an [`x25519_dalek::StaticSecret`].
+    ///
+    /// This conversion allows using the same underlying scalar from the Ed25519 secret key
+    /// for X25519 Diffie-Hellman key exchange. The returned `StaticSecret` can then be used
+    /// in key agreement protocols to establish a shared secret with another party’s
+    /// X25519 public key.
+    fn to_x25519(&self) -> x25519_dalek::StaticSecret {
+        let scalar_bytes = self.inner.to_scalar_bytes();
+        x25519_dalek::StaticSecret::from(scalar_bytes)
+    }
+}
+
+// PUBLIC KEY
+// ================================================================================================
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PublicKey {
+    pub(crate) inner: ed25519_dalek::VerifyingKey,
+}
+
+impl PublicKey {
+    /// Returns a commitment to the public key using the RPO256 hash function.
+    pub fn to_commitment(&self) -> Word {
+        <Self as SequentialCommit>::to_commitment(self)
+    }
+
+    /// Verifies a signature against this public key and message.
+    pub fn verify(&self, message: Word, signature: &Signature) -> bool {
+        let message_bytes: [u8; 32] = message.into();
+        self.inner.verify(&message_bytes, &signature.inner).is_ok()
+    }
+
+    /// Convert to a X25519 public key which can be used in a DH key exchange protocol.
+    pub(crate) fn to_x25519(&self) -> x25519_dalek::PublicKey {
+        let mont_point = self.inner.to_montgomery();
+        x25519_dalek::PublicKey::from(mont_point.to_bytes())
+    }
+}
+
+impl SequentialCommit for PublicKey {
+    type Commitment = Word;
+
+    fn to_elements(&self) -> Vec<Felt> {
+        bytes_to_elements_with_padding(&self.to_bytes())
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum PublicKeyError {
+    #[error("Could not verify with given public key and signature")]
+    VerificationFailed,
+}
+
+// SIGNATURE
+// ================================================================================================
+
+/// EdDSA (Ed25519) signature
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Signature {
+    inner: ed25519_dalek::Signature,
+}
+
+impl Signature {
+    /// Verify against (message, public key).
+    pub fn verify(&self, message: Word, pub_key: &PublicKey) -> bool {
+        pub_key.verify(message, self)
+    }
+}
+
+// SERIALIZATION / DESERIALIZATION
+// ================================================================================================
+
+impl Serializable for SecretKey {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        target.write_bytes(&self.inner.to_bytes());
+    }
+}
+
+impl Deserializable for SecretKey {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let bytes: [u8; SECRET_KEY_BYTES] = source.read_array()?;
```

Comment:
This may be a larger refactor but since we're using `Zeroize` in other places, it would be consistent to zeroize the bytes from the `source` as well as those in `bytes`